redwood-protocol-guest/src/commonMain/kotlin/app/cash/redwood/protocol/guest/DefaultProtocolState.kt

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) 2024 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.redwood.protocol.guest
+
+import app.cash.redwood.RedwoodCodegenApi
+import app.cash.redwood.protocol.Change
+import app.cash.redwood.protocol.ChangesSink
+import app.cash.redwood.protocol.ChildrenChange
+import app.cash.redwood.protocol.ChildrenTag
+import app.cash.redwood.protocol.Create
+import app.cash.redwood.protocol.Id
+import app.cash.redwood.protocol.ModifierChange
+import app.cash.redwood.protocol.ModifierElement
+import app.cash.redwood.protocol.PropertyChange
+import app.cash.redwood.protocol.PropertyTag
+import app.cash.redwood.protocol.RedwoodVersion
+import app.cash.redwood.protocol.WidgetTag
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonPrimitive
+
+/** @suppress For generated code use only. */
+@RedwoodCodegenApi
+public class DefaultProtocolState(
+  public override val json: Json = Json.Default,
+  hostVersion: RedwoodVersion,
+) : ProtocolState {
+  private var nextValue = Id.Root.value + 1
+  private val widgets = mutableMapOf<Int, ProtocolWidget>()
+  private var changes = mutableListOf<Change>()
+  private lateinit var changesSink: ChangesSink
+
+  public override val synthesizeSubtreeRemoval: Boolean = hostVersion < RedwoodVersion("0.10.0-SNAPSHOT")
+
+  override fun initChangesSink(changesSink: ChangesSink) {
+    this.changesSink = changesSink
+  }
```

Comment:
Eeek not my favorite. I mean I know we have to have a break in the cycle _somewhere_ for it to all work, but I prefer to pretend it doesn't exist.